sys/stat.c

```diff
@@ -24,62 +24,76 @@
 
 #include "posix/utils.h"
 
+
 /* path needs to be canonical */
 static int _stat_abs(const char *path, struct stat *buf)
 {
 	oid_t oid, dev;
 	msg_t msg = {0};
+	int err;
 
-	if (lookup(path, &oid, &dev) < 0) {
-		errno = ENOENT;
-		return -1;
-	}
+	do {
+		if (lookup(path, &oid, &dev) < 0) {
+			err = -ENOENT;
+			break;
+		}
 
-	memset(buf, 0, sizeof(struct stat));
+		memset(buf, 0, sizeof(struct stat));
 
-	buf->st_dev = oid.port;
-	buf->st_ino = oid.id;
-	buf->st_rdev = dev.port;
+		buf->st_dev = oid.port;
+		buf->st_ino = oid.id;
+		buf->st_rdev = dev.port;
 
-	msg.type = mtGetAttr;
-	msg.i.attr.oid = oid;
-	msg.i.attr.val = 0;
+		msg.type = mtGetAttr;
+		msg.i.attr.oid = oid;
 
-	msg.i.attr.type = atMTime;
-	if (!msgSend(oid.port, &msg))
+		msg.i.attr.type = atMTime;
+		if (((err = msgSend(oid.port, &msg)) < 0) || ((err = msg.o.attr.err) < 0))
+			break;
 		buf->st_mtime = msg.o.attr.val;
 
-	msg.i.attr.type = atATime;
-	if (!msgSend(oid.port, &msg))
+		msg.i.attr.type = atATime;
+		if (((err = msgSend(oid.port, &msg)) < 0) || ((err = msg.o.attr.err) < 0))
+			break;
 		buf->st_atime = msg.o.attr.val;
 
-	msg.i.attr.type = atCTime;
-	if (!msgSend(oid.port, &msg))
+		msg.i.attr.type = atCTime;
+		if (((err = msgSend(oid.port, &msg)) < 0) || ((err = msg.o.attr.err) < 0))
+			break;
 		buf->st_ctime = msg.o.attr.val;
 
-	msg.i.attr.type = atLinks;
-	if (!msgSend(oid.port, &msg))
+		msg.i.attr.type = atLinks;
+		if (((err = msgSend(oid.port, &msg)) < 0) || ((err = msg.o.attr.err) < 0))
+			break;
 		buf->st_nlink = msg.o.attr.val;
 
-	msg.i.attr.type = atMode;
-	if (!msgSend(oid.port, &msg))
+		msg.i.attr.type = atMode;
+		if (((err = msgSend(oid.port, &msg)) < 0) || ((err = msg.o.attr.err) < 0))
+			break;
 		buf->st_mode = msg.o.attr.val;
 
-	msg.i.attr.type = atUid;
-	if (!msgSend(oid.port, &msg))
+		msg.i.attr.type = atUid;
+		if (((err = msgSend(oid.port, &msg)) < 0) || ((err = msg.o.attr.err) < 0))
+			break;
 		buf->st_uid = msg.o.attr.val;
 
-	msg.i.attr.type = atGid;
-	if (!msgSend(oid.port, &msg))
+		msg.i.attr.type = atGid;
+		if (((err = msgSend(oid.port, &msg)) < 0) || ((err = msg.o.attr.err) < 0))
+			break;
 		buf->st_gid = msg.o.attr.val;
 
-	msg.i.attr.type = atSize;
-	if (!msgSend(oid.port, &msg))
+		msg.i.attr.type = atSize;
+		if (((err = msgSend(oid.port, &msg)) < 0) || ((err = msg.o.attr.err) < 0))
+			break;
 		buf->st_size = msg.o.attr.val;
 
-	return EOK;
+		err = EOK;
+	} while (0);
+
+	return SET_ERRNO(err);
```

Comment:
I think we should set ERRNO in `stat()`

Reply:
Changed as requested. ERRNO is set in `stat()` and `lstat()`.